File: packages/ts/src/validate.ts
/**
 * Optional validation through the Standard Schema interface — SPEC §5, kinds `validation` and
 * `unknown-key`.
 *
 * No validator is depended on. Zod, Valibot, and ArkType all expose `~standard`, so the caller
 * brings theirs and this package keeps three runtime dependencies.
 */

import { ConfigError } from './errors.js'
import type { StandardSchemaV1, WarningSink } from './types.js'

function issueKeyPath(issue: StandardSchemaV1.Issue): string | undefined {
  if (issue.path === undefined || issue.path.length === 0) return undefined
  return issue.path
    .map((segment) =>
      typeof segment === 'object' && segment !== null ? String(segment.key) : String(segment),
    )
    .join('.')
}

/**
 * Best effort at the schema's declared top-level keys, for `strict`. Standard Schema does not
 * expose them, so this reaches for the two shapes that cover most of the ecosystem and gives up
 * cleanly otherwise — see the README: where the validator does not expose its keys, `strict`
 * defers to the validator's own strict mode rather than silently doing nothing.
 */
function knownTopLevelKeys(schema: StandardSchemaV1): string[] | null {
  const candidate = schema as unknown as Record<string, unknown>
  const shape = candidate['shape'] ?? (candidate['_def'] as Record<string, unknown> | undefined)?.['shape']
  const resolved = typeof shape === 'function' ? (shape as () => unknown)() : shape
  if (resolved !== null && typeof resolved === 'object') return Object.keys(resolved)
  const entries = candidate['entries'] // Valibot
  if (entries !== null && typeof entries === 'object') return Object.keys(entries as object)
  return null
}

export interface ValidateOptions {
  strict?: boolean
  onWarning?: WarningSink
}

/**
 * Returns the validator's **output** value, not the input. A schema that coerces or fills in
 * defaults therefore changes what `load` returns, which is the useful behavior and is
 * documented as such in the README.
 */
export async function validateConfig(
  config: Record<string, unknown>,
  schema: StandardSchemaV1,
  options: ValidateOptions = {},
): Promise<unknown> {
  checkUnknownKeys(config, schema, options)
  const result = await schema['~standard'].validate(config)
  return unwrap(result)
}

export function validateConfigSync(
  config: Record<string, unknown>,
  schema: StandardSchemaV1,
  options: ValidateOptions = {},
): unknown {
  checkUnknownKeys(config, schema, options)
  const result = schema['~standard'].validate(config)
  if (result instanceof Promise) {
    throw new ConfigError('validation', 'this schema validates asynchronously; use load() rather than loadSync()')
  }
  return unwrap(result)
}

function unwrap(result: StandardSchemaV1.Result<unknown>): unknown {
  if (result.issues === undefined) return result.value

  const [first] = result.issues
  const keyPath = first === undefined ? undefined : issueKeyPath(first)
  // The first issue by key path is the one named, but all of them ride along on `issues`: a
  // config with four wrong values should not take four runs to fix.
  throw new ConfigError('validation', first?.message ?? 'configuration failed validation', {
    ...(keyPath === undefined ? {} : { keyPath }),
    issues: result.issues,
  })
}

function checkUnknownKeys(
  config: Record<string, unknown>,
  schema: StandardSchemaV1,
  options: ValidateOptions,
): void {
  const known = knownTopLevelKeys(schema)
  if (known === null) return
  const unknown = Object.keys(config).filter((key) => !known.includes(key))
  if (unknown.length === 0) return

  const list = unknown.join(', ')
  if (options.strict === true) {
    throw new ConfigError('unknown-key', `configuration has unknown keys: ${list}`, {
      keyPath: unknown[0] as string,
    })
  }
  options.onWarning?.(`configuration has unknown keys: ${list} (set strict to make this an error)`)
}

File: packages/py/src/config_discovery/validate.py
"""Optional validation - SPEC section 5, kinds ``validation`` and ``unknown-key``.

No validator is depended on. The caller passes a callable, so a Pydantic model, an attrs class,
or a hand-written function all work unchanged.
"""

from __future__ import annotations

import inspect
from typing import Any, Callable

from .errors import ConfigError


def _known_fields(schema: Any) -> list[str] | None:
    """Best effort at the schema's declared top-level fields, for ``strict``.

    A plain callable declares nothing, so this gives up cleanly rather than pretending - see the
    README: where the fields are not discoverable, ``strict`` defers to the validator's own
    strictness rather than silently doing nothing.
    """
    model_fields = getattr(schema, "model_fields", None)  # Pydantic v2
    if isinstance(model_fields, dict):
        return list(model_fields)
    dataclass_fields = getattr(schema, "__dataclass_fields__", None)
    if isinstance(dataclass_fields, dict):
        return list(dataclass_fields)
    # A function's __annotations__ are its *parameters*, not a set of fields. Reading them as
    # fields makes every plain-callable schema reject every key it was handed.
    if inspect.isroutine(schema):
        return None
    annotations = getattr(schema, "__annotations__", None)
    if isinstance(annotations, dict) and annotations:
        return [name for name in annotations if name != "return"]
    return None


def _issues_of(error: BaseException) -> list[Any] | None:
    """Pydantic's structured errors, if this is a Pydantic error.

    Imported lazily and defensively so the library keeps no dependency on Pydantic - and so a
    validator that merely *has* an ``errors`` attribute cannot crash the error path.
    """
    errors = getattr(error, "errors", None)
    if not callable(errors):
        return None
    try:
        result = errors()
    except Exception:  # noqa: BLE001 - a validator's own failure must not mask the real error
        return None
    return list(result) if isinstance(result, (list, tuple)) else None


def _key_path_of(issues: list[Any] | None) -> str | None:
    if not issues:
        return None
    first = issues[0]
    location = first.get("loc") if isinstance(first, dict) else None
    if not location:
        return None
    return ".".join(str(segment) for segment in location)


def apply_schema(
    config: dict[str, Any],
    schema: Callable[[dict[str, Any]], Any],
    *,
    strict: bool = False,
    on_warning: Callable[[str], None] | None = None,
) -> Any:
    """Validate *config* and return **the validator's output**.

    Returning the output rather than the input is deliberate: a schema that coerces values or
    fills in defaults therefore changes what ``load`` returns, which is the useful behavior.
    """
    known = _known_fields(schema)
    if known is not None:
        unknown = [key for key in config if key not in known]
        if unknown and strict:
            raise ConfigError(
                "unknown-key",
                f"configuration has unknown keys: {', '.join(unknown)}",
                key_path=unknown[0],
            )
        if unknown and on_warning is not None:
            on_warning(
                f"configuration has unknown keys: {', '.join(unknown)} "
                "(pass strict=True to make this an error)"
            )

    # A Pydantic model class is not callable with a positional dict - BaseModel.__init__ takes
    # keywords - so the one accommodation this module makes is to prefer `model_validate` when
    # the schema offers it. Everything else is called directly, as a plain callable.
    validate = getattr(schema, "model_validate", None)
    if not callable(validate):
        validate = schema

    try:
        return validate(config)
    except ConfigError:
        raise
    except Exception as error:  # noqa: BLE001 - any validator failure is one `kind`
        issues = _issues_of(error)
        raise ConfigError(
            "validation",
            str(error),
            key_path=_key_path_of(issues),
            issues=issues,
        ) from error
